lambdas/drivingpermitcheck/src/main/java/uk/gov/di/ipv/cri/drivingpermit/api/service/dva/DvaThirdPartyDocumentGateway.java

```diff
@@ -1,41 +1,286 @@
+        }
+    }
+
+    private DocumentCheckResult responseHandler(CloseableHttpResponse httpResponse)
+            throws IOException, ParseException, JOSEException, CertificateException,
+                    OAuthHttpResponseExceptionWithErrorBody {
+        int statusCode = httpResponse.getStatusLine().getStatusCode();
+
+        HttpEntity entity = httpResponse.getEntity();
+        String responseBody = EntityUtils.toString(entity);
+
+        LOGGER.info("Third party response code {}", statusCode);
+
+        if (statusCode == 200) {
+            try {
+                if (configurationService.isLogThirdPartyResponse()) {
+                    LOGGER.info("DVA response " + responseBody);
+                }
+                DvaResponse unwrappedDvaResponse =
+                        dvaCryptographyService.unwrapDvaResponse(responseBody);
+                validateDvaResponse(unwrappedDvaResponse);
+
+                LOGGER.info("Third party response successfully mapped");
+                // Might be worth making a new metric for dvaDirect?
+                eventProbe.counterMetric(THIRD_PARTY_DVA_RESPONSE_OK);
+
+                DocumentCheckResult documentCheckResult = new DocumentCheckResult();
+                documentCheckResult.setExecutedSuccessfully(true);
+                documentCheckResult.setValid(unwrappedDvaResponse.isValidDocument());
+
+                return documentCheckResult;
+            } catch (IpvCryptoException e) {
+                // Seen when a signing cert has expired and all message signatures fail verification
+                // We need to log this specific error message from the IpvCryptoException for
+                // context
+                LOGGER.error(e.getMessage(), e);
+                eventProbe.counterMetric(THIRD_PARTY_DVA_RESPONSE_TYPE_ERROR);
+                throw new OAuthHttpResponseExceptionWithErrorBody(
+                        HttpStatusCode.INTERNAL_SERVER_ERROR,
+                        ErrorResponse.FAILED_TO_UNWRAP_DVA_RESPONSE);
+                // may need to make new mappings for DVA responseCodes
+            }
+        } else {
+
+            String responseText = responseBody == null ? "No Text Found" : responseBody;
+
+            LOGGER.error(
+                    "DVA replied with HTTP status code {}, response text: {}",
+                    statusCode,
+                    responseText);
+
+            eventProbe.counterMetric(THIRD_PARTY_DVA_RESPONSE_TYPE_ERROR);
+
+            if (statusCode >= 300 && statusCode <= 399) {
+                // Not Seen
+                throw new OAuthHttpResponseExceptionWithErrorBody(
+                        HttpStatusCode.INTERNAL_SERVER_ERROR,
+                        ErrorResponse.DVA_ERROR_HTTP_30X); // need DVAErrorResponse
+            } else if (statusCode >= 400 && statusCode <= 499) {
+                if (statusCode == 400) {
+                    LOGGER.info("DVA replied with an InvalidRequestError");
```

Comment:
This should probably be a warn.
Also we could add something like 
`DVA replied with an InvalidRequestError. Please check the schema and request sent to DVA`